pkg/vault/contracts/AggregatorBatchRouter.sol

```diff
@@ -0,0 +1,476 @@
+        for (uint256 i = 0; i < tokensOut.length; ++i) {
+            amountsOut[i] = _currentSwapTokenOutAmounts().tGet(tokensOut[i]);
+        }
+    }
+
+    function _computePathAmountsOut(
+        SwapExactInHookParams calldata params
+    ) internal returns (uint256[] memory pathAmountsOut) {
+        pathAmountsOut = new uint256[](params.paths.length);
+
+        if (EVMCallModeHelpers.isStaticCall() == false) {
+            for (uint256 i = 0; i < params.paths.length; ++i) {
+                SwapPathExactAmountIn memory path = params.paths[i];
+                _currentSwapTokensIn().add(address(path.tokenIn));
+                _currentSwapTokenInAmounts().tAdd(address(path.tokenIn), path.exactAmountIn);
+            }
+
+            address[] memory tokensIn = _currentSwapTokensIn().values();
+            for (uint256 i = 0; i < tokensIn.length; ++i) {
+                uint256 amount = _currentSwapTokenInAmounts().tGet(tokensIn[i]);
+                uint256 tokenInCredit = _vault.settle(IERC20(tokensIn[i]), amount);
+                if (tokenInCredit < amount) {
+                    revert InsufficientPayment();
+                }
+
+                _currentSwapTokenInAmounts().tSet(tokensIn[i], 0);
+            }
+        }
+
+        for (uint256 i = 0; i < params.paths.length; ++i) {
+            SwapPathExactAmountIn memory path = params.paths[i];
+
+            // These two variables shall be updated at the end of each step to be used as inputs of the next one.
+            // The initial values are the given token and amount in for the current path.
+            uint256 stepExactAmountIn = path.exactAmountIn;
+            IERC20 stepTokenIn = path.tokenIn;
+
+            for (uint256 j = 0; j < path.steps.length; ++j) {
+                SwapStepLocals memory stepLocals;
+                stepLocals.isLastStep = (j == path.steps.length - 1);
+                stepLocals.isFirstStep = (j == 0);
+                uint256 minAmountOut;
+                uint256 amountOut;
+
+                // minAmountOut only applies to the last step.
+                if (stepLocals.isLastStep) {
+                    minAmountOut = path.minAmountOut;
+                } else {
+                    minAmountOut = 0;
+                }
+
+                SwapPathStep memory step = path.steps[j];
+
+                if (step.isBuffer) {
+                    (, , amountOut) = _vault.erc4626BufferWrapOrUnwrap(
+                        BufferWrapOrUnwrapParams({
+                            kind: SwapKind.EXACT_IN,
+                            direction: step.pool == address(stepTokenIn)
+                                ? WrappingDirection.UNWRAP
+                                : WrappingDirection.WRAP,
+                            wrappedToken: IERC4626(step.pool),
+                            amountGivenRaw: stepExactAmountIn,
+                            limitRaw: minAmountOut
+                        })
+                    );
+                } else {
```

Comment:
Maybe we should require that `step.pool != stepTokenIn` and `step.pool != step.tokenOut` if it's not a buffer operation? Those are really not supported and will revert when calling `swap` with a not-so-nice error.